extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreIntegrationTest.java

```diff
@@ -0,0 +1,129 @@
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static com.hazelcast.mapstore.GenericMapStore.EXTERNAL_REF_ID_PROPERTY;
+import static com.hazelcast.mapstore.GenericMapStore.TYPE_NAME_PROPERTY;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.util.Lists.newArrayList;
+
+public class GenericMapStoreIntegrationTest extends JdbcSqlTestSupport {
+
+    private String tableName;
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        databaseProvider = new H2DatabaseProvider();
+        dbConnectionUrl = databaseProvider.createDatabase(JdbcSqlTestSupport.class.getName());
+
+        Config config = smallInstanceConfig();
+        config.getSerializationConfig().getCompactSerializationConfig().setEnabled(true);
+        // Need to set filtering class loader so the members don't deserialize into class but into GenericRecord
+        config.setClassLoader(new FilteringClassLoader(newArrayList("org.example"), null));
+
+        config.addExternalDataStoreConfig(
+                new ExternalDataStoreConfig(TEST_DATABASE_REF)
+                        .setClassName(JdbcDataStoreFactory.class.getName())
+                        .setProperty("jdbcUrl", dbConnectionUrl)
+        );
+
+
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getSerializationConfig().getCompactSerializationConfig().setEnabled(true);
+
+        initializeWithClient(2, config, clientConfig);
+        sqlService = instance().getSql();
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        tableName = randomTableName();
+        createTable(tableName);
+        insertItems(tableName, 1);
+
+        MapConfig mapConfig = new MapConfig(tableName);
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setClassName(GenericMapStore.class.getName());
+        mapStoreConfig.setProperty(EXTERNAL_REF_ID_PROPERTY, TEST_DATABASE_REF);
+        mapStoreConfig.setProperty(TYPE_NAME_PROPERTY, "org.example.Person");
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        instance().getConfig().addMapConfig(mapConfig);
+    }
+
+    @Test
+    public void testGet() {
+        HazelcastInstance client = client();
+        IMap<Integer, Person> map = client.getMap(tableName);
+
+        Person p = map.get(0);
+        assertThat(p.getId()).isEqualTo(0);
+        assertThat(p.getName()).isEqualTo("name-0");
+    }
+
+    @Test
+    public void testPut() {
+        HazelcastInstance client = client();
+        IMap<Integer, Person> map = client.getMap(tableName);
+
+        map.put(42, new Person(42, "name-42"));
+
+        assertThat(jdbcRows(tableName)).containsOnly(
+                new Row(0, "name-0"),
+                new Row(42, "name-42")
+        );
+    }
+
+    @Test
+    public void testPutWhenExists() {
+        HazelcastInstance client = client();
+        IMap<Integer, Person> map = client.getMap(tableName);
```

Comment:
we could pre-check that we already have  a record with `id=0`

Reply:
458b3730